src/python/pants/backend/python/rules/run_setup_py.py

```diff
@@ -178,14 +181,84 @@ class SetupPyChrootRequest:
     py2: bool  # Whether to use py2 or py3 package semantics.
 
 
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class SetupKwargs:
+    """The keyword arguments to the `setup()` function in the generated `setup.py`."""
+
+    json_str: str
+
+    def __init__(
+        self, kwargs: Mapping[str, Any], *, address: Address, _allow_banned_keys: bool = False
+    ) -> None:
+        super().__init__()
+        if "version" not in kwargs:
+            raise ValueError(f"Missing a `version` kwarg in the `provides` field for {address}.")
+
+        if not _allow_banned_keys:
+            for arg in {"data_files", "package_dir", "package_data", "packages"}:
+                if arg in kwargs:
+                    raise ValueError(
+                        f"{arg} cannot be set in the `provides` field for {address}, but it was "
+                        f"set to {kwargs[arg]}. Pants will dynamically set the value for you."
+                    )
+
+        # We convert to a `str` so that this type can be hashable. We don't use `FrozenDict`
+        # because it would require that all values are immutable, and we may have lists and
+        # dictionaries as values. It's too difficult/clunky to convert those all, then to convert
+        # them back out of `FrozenDict`.
+        self.json_str = json.dumps(kwargs, sort_keys=True)
+
+    @memoized_property
+    def kwargs(self) -> Dict[str, Any]:
+        return cast(Dict[str, Any], json.loads(self.json_str))
+
+    @property
+    def name(self) -> str:
+        return cast(str, self.kwargs["name"])
+
+    @property
+    def version(self) -> str:
+        return cast(str, self.kwargs["version"])
+
+
+@union
+@dataclass(frozen=True)  # type: ignore[misc]
+class SetupKwargsPluginRequest(ABC):
+    """An entry point to customize the kwargs used for the `setup()` function, rather than the
+    default behavior of using what was explicitly provided in the BUILD file.
+
+    To add a plugin, subclass `SetupKwargsPluginRequest`, register the rule
+    `UnionRule(SetupKwargsPluginRequest, MyCustomSetupKwargsPluginRequest)`, and add a rule that
+    takes your subclass as a parameter and returns `SetupKwargs`.
+    """
+
+    target: Target
+
+    @classmethod
+    @abstractmethod
+    def is_valid(cls, target: Target) -> bool:
+        """Whether the kwargs implementation should be used for this target or not."""
+
+    @property
+    def explicit_kwargs(self) -> Dict[str, Any]:
+        return self.target[PythonProvidesField].value.kwargs
+
+
+class FinalizedSetupKwargs(SetupKwargs):
+    """The final kwargs used for the `setup()` function, after Pants added requirements and sources
+    information."""
+
+    def __init__(self, kwargs: Mapping[str, Any], *, address: Address) -> None:
+        super().__init__(kwargs, address=address, _allow_banned_keys=True)
```

Comment:
This took me a while to figure out. Normally, we want to validate that certain kwargs are not set by the user or plugin. But in the final thing, we do want those set.

This was the best design I could come up with to express that relationship via types.

Reply:
Did you look at doing this via containment, not subclassing?

Reply:
Yes. I originally tried that, but it resulted in lots of duplication like having to duplicate the whole `json_str` thing to be able to store a dictionary in a frozen dataclass.

This seemed like the least bad out of multiple bad options.

Reply:
What I was thinking was that the `json_str` thing can be contained inside two unrelated classes, one that does the check and one that doesn't.

Reply:
So it's almost like a "serialized JSON" field type, in practice.

Reply:
Do you mean something like this?

```diff
diff --git a/src/python/pants/backend/python/rules/run_setup_py.py b/src/python/pants/backend/python/rules/run_setup_py.py
index 349f72125..e0d502264 100644
--- a/src/python/pants/backend/python/rules/run_setup_py.py
+++ b/src/python/pants/backend/python/rules/run_setup_py.py
@@ -183,26 +183,16 @@ class SetupPyChrootRequest:
 
 @frozen_after_init
 @dataclass(unsafe_hash=True)
-class SetupKwargs:
-    """The keyword arguments to the `setup()` function in the generated `setup.py`."""
-
+class _SetupKwargsJson:
     json_str: str
 
     def __init__(
-        self, kwargs: Mapping[str, Any], *, address: Address, _allow_banned_keys: bool = False
+        self, kwargs: Mapping[str, Any], *, address: Address
     ) -> None:
         super().__init__()
         if "version" not in kwargs:
             raise ValueError(f"Missing a `version` kwarg in the `provides` field for {address}.")
 
-        if not _allow_banned_keys:
-            for arg in {"data_files", "package_dir", "package_data", "packages"}:
-                if arg in kwargs:
-                    raise ValueError(
-                        f"{arg} cannot be set in the `provides` field for {address}, but it was "
-                        f"set to {kwargs[arg]}. Pants will dynamically set the value for you."
-                    )
-
         # We convert to a `str` so that this type can be hashable. We don't use `FrozenDict`
         # because it would require that all values are immutable, and we may have lists and
         # dictionaries as values. It's too difficult/clunky to convert those all, then to convert
@@ -222,6 +212,38 @@ class SetupKwargs:
         return cast(str, self.kwargs["version"])
 
 
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class SetupKwargs:
+    """The keyword arguments to the `setup()` function in the generated `setup.py`."""
+
+    _json: _SetupKwargsJson
+
+    def __init__(
+        self, kwargs: Mapping[str, Any], *, address: Address
+    ) -> None:
+        super().__init__()
+        for arg in {"data_files", "package_dir", "package_data", "packages"}:
+            if arg in kwargs:
+                raise ValueError(
+                    f"{arg} cannot be set in the `provides` field for {address}, but it was "
+                    f"set to {kwargs[arg]}. Pants will dynamically set the value for you."
+                )
+        self._json = _SetupKwargsJson(kwargs, address=address)
+
+    @property
+    def kwargs(self) -> Dict[str, Any]:
+        return self._json.kwargs
+
+    @property
+    def name(self) -> str:
+        return self._json.name
+
+    @property
+    def version(self) -> str:
+        return self._json.version
+
+
 @union
 @dataclass(frozen=True)  # type: ignore[misc]
 class SetupKwargsPluginRequest(ABC):
@@ -245,12 +267,29 @@ class SetupKwargsPluginRequest(ABC):
         return self.target[PythonProvidesField].value.kwargs
 
 
-class FinalizedSetupKwargs(SetupKwargs):
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class FinalizedSetupKwargs:
     """The final kwargs used for the `setup()` function, after Pants added requirements and sources
     information."""
 
+    _json: _SetupKwargsJson
+
     def __init__(self, kwargs: Mapping[str, Any], *, address: Address) -> None:
-        super().__init__(kwargs, address=address, _allow_banned_keys=True)
+        super().__init__()
+        self._json = _SetupKwargsJson(kwargs, address=address)
+
+    @property
+    def kwargs(self) -> Dict[str, Any]:
+        return self._json.kwargs
+
+    @property
+    def name(self) -> str:
+        return self._json.name
+
+    @property
+    def version(self) -> str:
+        return self._json.version
 
 
 @dataclass(frozen=True)
```

This seems much worse, imo.